File: apps/android/app/src/main/kotlin/com/loonext/android/features/settings/HoursSection.kt
package com.loonext.android.features.settings

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.loonext.android.features.compose.usSendApproved
import com.loonext.android.features.compose.usTextingOff
import com.loonext.android.core.i18n.LocalAppLocale
import com.loonext.android.core.i18n.t
import com.loonext.android.core.model.CompanyView
import com.loonext.android.core.model.DayHours
import com.loonext.android.ui.common.userMessage
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject

/**
 * One weekday row's editable state.
 *
 * `internal`, not private: #307 gives a single NUMBER its own week, and a
 * second copy of this would have drifted from the workspace's the first time
 * either was touched.
 */
internal data class DayForm(
    val weekday: String,
    val enabled: Boolean,
    val open: String,
    val close: String,
)

internal fun toFormState(hours: Map<String, DayHours?>): List<DayForm> =
    WEEKDAY_KEYS.map { key ->
        val window = hours[key]
        DayForm(
            weekday = key,
            enabled = window != null,
            open = window?.open ?: "09:00",
            close = window?.close ?: "17:00",
        )
    }

/**
 * Business hours & away reply (#157): the per-weekday open/close grid with
 * enable switches, and the after-hours auto-reply with merge fields and a live
 * preview that matches the wire byte-for-byte.
 */
@Composable
fun HoursSection(
    scope: SettingsScope,
    company: CompanyView,
    onCompanyUpdated: (CompanyView) -> Unit,
) {
    BusinessHoursCard(scope, company, onCompanyUpdated)
    // #402: directly under the weekly schedule it overrides. These dates only
    // mean anything as an exception to it, and an owner looking for "we're
    // shut on Boxing Day" looks where they set their hours.
    ClosedDatesCard(scope, company, onCompanyUpdated)
    AwayReplyCard(scope, company, onCompanyUpdated)
    // #460: directly beneath the away message, which is the sentence that
    // TELLS a customer the word. An owner changing the word has to see the
    // offer in the same scroll.
    EmergencyCard(scope, company, onCompanyUpdated)
    // #237: last, and on this screen rather than a section of its own. Every
    // card above answers "what do we send automatically, and in whose words" —
    // the away reply, the closed dates, the emergency reply. A reminder is the
    // same question with a different trigger, and the settings list is already
    // long enough that a two-rule form does not earn another row.
    ReminderRulesCard(scope)
    // #244: with the reminder card and everything above it — every card on
    // this screen answers "what happens outside working hours". The away reply
    // is what the CUSTOMER gets; this is who on the crew is woken.
    OnCallCard(scope)
}

@Composable
private fun BusinessHoursCard(
    scope: SettingsScope,
    company: CompanyView,
    onCompanyUpdated: (CompanyView) -> Unit,
) {
    val canEdit = SettingsRoleGate.canEditWorkspace(scope.role)
    val initial = remember(company.business_hours) { toFormState(company.business_hours) }
    var days by remember(initial) { mutableStateOf(initial) }
    var saving by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    val coroutines = rememberCoroutineScope()
    // #228: the save failure is written from a coroutine, outside composition.
    val locale = LocalAppLocale.current

    val dirty = days != initial
    val allValid = days.all { !it.enabled || isValidDayWindow(it.open, it.close) }
    // Reported from the coroutine the Save press starts, so it is read here.
    val savedMessage = t("settings.hoursSaved")

    fun patchDay(weekday: String, transform: (DayForm) -> DayForm) {
        days = days.map { if (it.weekday == weekday) transform(it) else it }
    }

    SettingsCard(
        title = t("settings.hoursTitle"),
        description = t(
            "settings.hoursIntro",
            "timezone" to company.timezone.replace('_', ' '),
        ),
    ) {
        days.forEach { day ->
            Row(
                Modifier
                    .fillMaxWidth()
                    .padding(vertical = 6.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                WeekdayRow(
                    day = day,
                    enabled = canEdit && !saving,
                    onChange = { updated -> patchDay(day.weekday) { updated } },
                )
            }
        }
        if (!allValid) {
            ReadOnlyLine(t("settings.hoursInvalid"))
        }
        InlineError(error)
        if (canEdit) {
            if (dirty) {
                Button(
                    onClick = {
                        error = null
                        saving = true
                        coroutines.launch {
                            try {
                                val body = buildJsonObject {
                                    putJsonObject("business_hours") {
                                        days.forEach { day ->
                                            if (day.enabled) {
                                                putJsonObject(day.weekday) {
                                                    put("open", day.open)
                                                    put("close", day.close)
                                                }
                                            }
                                        }
                                    }
                                }
                                val updated = scope.repo.updateCompany(scope.companyId, body)
                                onCompanyUpdated(updated)
                                scope.showMessage(savedMessage)
                            } catch (cause: Exception) {
                                error = cause.userMessage(locale)
                            } finally {
                                saving = false
                            }
                        }
                    },
                    enabled = allValid && !saving,
                    modifier = Modifier.padding(top = 10.dp),
                ) { Text(if (saving) t("common.saving") else t("settings.hoursSaveAction")) }
            }
        } else {
            Spacer(Modifier.height(4.dp))
            ReadOnlyLine(t("settings.hoursReadOnly"))
        }
    }
}

@Composable
private fun TimeField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    enabled: Boolean,
    modifier: Modifier = Modifier,
) {
    OutlinedTextField(
        value = value,
        onValueChange = { next -> if (next.length <= 5) onValueChange(next) },
        modifier = modifier,
        singleLine = true,
        enabled = enabled,
        isError = !isValidHhmm(value),
        label = { Text(label, style = MaterialTheme.typography.labelSmall) },
        placeholder = { Text("09:00") },
        textStyle = MaterialTheme.typography.bodyMedium,
    )
}

@Composable
private fun AwayReplyCard(
    scope: SettingsScope,
    company: CompanyView,
    onCompanyUpdated: (CompanyView) -> Unit,
) {
    val canEdit = SettingsRoleGate.canEditWorkspace(scope.role)
    var enabled by remember(company.away_enabled) { mutableStateOf(company.away_enabled) }
    var message by remember(company.away_message) {
        mutableStateOf(company.away_message.orEmpty())
    }
    var emergency by remember(company.emergency_keyword_enabled) {
        mutableStateOf(company.emergency_keyword_enabled)
    }
    var saving by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    // #228: the save failure is written from a coroutine, outside composition.
    val locale = LocalAppLocale.current
    val coroutines = rememberCoroutineScope()

    val trimmed = message.trim()
    val dirty = enabled != company.away_enabled ||
        trimmed != company.away_message.orEmpty().trim() ||
        emergency != company.emergency_keyword_enabled
    // What actually goes out — the owner's text if they wrote one, else the
    // product default. The preview and the #414 emergency check both read
    // THIS, so the screen can never approve of a message that isn't sending.
    // #414 ask 5: the SERVER says what will actually send.
    val effectiveMessage = trimmed.ifEmpty { company.away_effective_message }
    // The preview reuses the wire's drop-empty semantics: {first_name} resolves
    // to a sample name here because the away reply DOES carry the contact.
    val preview = applyMergeFields(
        text = effectiveMessage,
        contactName = SAMPLE_FIRST_NAME,
        businessName = company.name,
    )
    val emergencyNotice = awayEmergencyNotice(
        emergencyEnabled = emergency,
        awayMessage = effectiveMessage,
        locale = locale,
        // #460: THIS workspace's words, resolved by the server. Warning against
        // the product list when the owner watches for their own would be the
        // product arguing with a setting it offers.
        keywords = company.effectiveEmergencyWords,
    )

    SettingsCard(
        title = t("settings.awayTitle"),
        description = t("settings.awayIntro"),
    ) {
        LabeledSwitchRow(
            label = t("settings.awayEnable"),
            supporting = t("settings.awayEnableHelp"),
            checked = enabled,
            onCheckedChange = { enabled = it },
            enabled = canEdit && !saving,
        )
        // The send gates refuse a US destination until the campaign is
        // approved, and the away reply is best-effort: a refusal is swallowed
        // so it never breaks inbound ingest. A switch reading ON while every
        // US customer gets silence is the first week of every US workspace.
        if (enabled && !usSendApproved(company)) {
            ReachNote(
                if (usTextingOff(company)) {
                    t("settings.awayUsTextingOff")
                } else {
                    t("settings.awayUsPending")
                },
            )
        }
        if (canEdit) {
            OutlinedTextField(
                value = message,
                onValueChange = { if (it.length <= 1000) message = it },
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 6.dp),
                minLines = 3,
                enabled = !saving,
                placeholder = { Text(company.away_effective_message) },
                // `{first_name}` and `{business_name}` survive the catalogue's
                // interpolation untouched — an unknown token is left visible on
                // purpose — and they are NOT translated in either language: they
                // are the literal merge fields an owner types into the box.
                supportingText = {
                    Text(t("settings.awayCount", "count" to message.length.toString()))
                },
            )
        }
        // #414: the switch sits with the message that makes the offer, not on
        // a separate notifications screen. They are one decision — a message
        // inviting URGENT with the mechanism off is the exact defect this
        // issue is about, and an owner can only see it if both are together.
        LabeledSwitchRow(
            label = t("settings.awayEmergencySwitch"),
            // #460: names the words THIS workspace watches for. Hardcoding the
            // product's four was fine until an owner could change them, at
            // which point a switch naming words nothing matches is the #414
            // defect in a different place.
            supporting = t(
                "settings.awayEmergencySwitchHelp",
                "words" to emergencyWordList(company.effectiveEmergencyWords, locale),
            ),
            checked = emergency,
            onCheckedChange = { emergency = it },
            enabled = canEdit && !saving,
        )
        // #453: which sentence appears is decided in SettingsLogic, mirroring
        // shared, so this screen, web and iOS cannot drift into three wordings
        // of the same warning. Only the tone-to-colour mapping is ours.
        emergencyNotice?.let { notice ->
            ReachNote(
                notice.text,
                tone = when (notice.tone) {
                    AwayNoticeTone.Warn -> NoteTone.Warn
                    AwayNoticeTone.Hint -> NoteTone.Neutral
                },
            )
        }
        PreviewBubble(label = t("settings.awayPreviewLabel"), text = preview)
        InlineError(error)
        if (canEdit) {
            if (dirty) {
                val needsMessage = t("settings.awayNeedsMessage")
                val awaySaved = t("settings.awaySaved")
                Button(
                    onClick = {
                        if (enabled && trimmed.isEmpty()) {
                            error = needsMessage
                            return@Button
                        }
                        error = null
                        saving = true
                        coroutines.launch {
                            try {
                                val body = buildJsonObject {
                                    put("away_enabled", enabled)
                                    if (trimmed.isEmpty()) put("away_message", JsonNull)
                                    else put("away_message", trimmed)
                                    put("emergency_keyword_enabled", emergency)
                                }
                                val updated = scope.repo.updateCompany(scope.companyId, body)
                                onCompanyUpdated(updated)
                                scope.showMessage(awaySaved)
                            } catch (cause: Exception) {
                                error = cause.userMessage(locale)
                            } finally {
                                saving = false
                            }
                        }
                    },
                    enabled = !saving,
                    modifier = Modifier.padding(top = 10.dp),
                ) { Text(if (saving) t("common.saving") else t("settings.awaySaveAction")) }
            }
        } else {
            Spacer(Modifier.height(4.dp))
            ReadOnlyLine(t("settings.awayReadOnly"))
        }
    }
}

/** A quiet message-bubble preview: exactly what the customer receives. */
@Composable
fun PreviewBubble(label: String, text: String, modifier: Modifier = Modifier) {
    Column(modifier.padding(top = 10.dp)) {
        Text(
            label,
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Spacer(Modifier.height(4.dp))
        Text(
            text,
            style = MaterialTheme.typography.bodyMedium,
            modifier = Modifier
                .fillMaxWidth()
                .background(
                    MaterialTheme.colorScheme.surfaceContainerHigh,
                    RoundedCornerShape(12.dp),
                )
                .padding(12.dp),
        )
    }
}

/**
 * One weekday, shared by the workspace's hours and a single line's (#307).
 *
 * Stateless on purpose: the caller owns the list and decides what saving
 * means, because the workspace saves through the company route and a line
 * saves through its own. This only knows how to show a day.
 */
@Composable
internal fun RowScope.WeekdayRow(
    day: DayForm,
    enabled: Boolean,
    onChange: (DayForm) -> Unit,
) {
    Switch(
        checked = day.enabled,
        onCheckedChange = { onChange(day.copy(enabled = it)) },
        enabled = enabled,
    )
    Spacer(Modifier.width(10.dp))
    Text(
        WEEKDAY_LABELS[day.weekday] ?: day.weekday,
        style = MaterialTheme.typography.bodyMedium,
        modifier = Modifier.width(86.dp),
    )
    if (day.enabled) {
        TimeField(
            value = day.open,
            onValueChange = { onChange(day.copy(open = it)) },
            label = t("settings.hoursOpen"),
            enabled = enabled,
            modifier = Modifier.weight(1f),
        )
        Text(
            t("settings.hoursTo"),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(horizontal = 6.dp),
        )
        TimeField(
            value = day.close,
            onValueChange = { onChange(day.copy(close = it)) },
            label = t("settings.hoursClose"),
            enabled = enabled,
            modifier = Modifier.weight(1f),
        )
    } else {
        Text(
            t("settings.hoursClosed"),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

File: apps/ios/Loonext/Features/Settings/HoursSection.swift
import SwiftUI


/// One weekday row's editable state.
struct DayForm: Equatable, Sendable {
    let weekday: String
    var enabled: Bool
    var open: String
    var close: String
}

func toFormState(_ hours: [String: DayHours?]) -> [DayForm] {
    weekdayKeys.map { key in
        let window = hours[key] ?? nil
        return DayForm(
            weekday: key,
            enabled: window != nil,
            open: window?.open ?? "09:00",
            close: window?.close ?? "17:00"
        )
    }
}

/// Business hours & away reply (#163): the per-weekday open/close grid with
/// enable switches, and the after-hours auto-reply with merge fields and a
/// live preview that matches the wire byte-for-byte.
@MainActor
struct HoursSectionView: View {
    let scope: SettingsScope
    let company: CompanyView
    let onCompanyUpdated: @MainActor (CompanyView) -> Void

    var body: some View {
        BusinessHoursCard(scope: scope, company: company, onCompanyUpdated: onCompanyUpdated)
        AwayReplyCard(scope: scope, company: company, onCompanyUpdated: onCompanyUpdated)
    }
}

// MARK: - Business hours

private struct BusinessHoursCard: View {
    let scope: SettingsScope
    let company: CompanyView
    let onCompanyUpdated: @MainActor (CompanyView) -> Void

    @State private var days: [DayForm]
    @State private var saving = false
    @State private var error: String?

    init(scope: SettingsScope, company: CompanyView, onCompanyUpdated: @escaping @MainActor (CompanyView) -> Void) {
        self.scope = scope
        self.company = company
        self.onCompanyUpdated = onCompanyUpdated
        _days = State(initialValue: toFormState(company.business_hours))
    }

    private var canEdit: Bool { SettingsRoleGate.canEditWorkspace(scope.role) }
    private var dirty: Bool { days != toFormState(company.business_hours) }
    private var allValid: Bool {
        days.allSatisfy { !$0.enabled || isValidDayWindow(open: $0.open, close: $0.close) }
    }

    var body: some View {
        SettingsCard(
            title: "Business hours",
            description: "When you're open, in \(company.timezone.replacingOccurrences(of: "_", with: " ")). "
                + "Texts that arrive outside these hours can get your away reply. This is "
                + "separate from each customer's texting quiet hours."
        ) {
            ForEach($days, id: \.weekday) { $day in
                HStack(spacing: 10) {
                    Toggle("", isOn: $day.enabled)
                        .labelsHidden()
                        .tint(BrandColor.olive)
                        .disabled(!canEdit || saving)
                    Text(weekdayLabels[day.weekday] ?? day.weekday)
                        .font(.callout)
                        .frame(width: 86, alignment: .leading)
                    if day.enabled {
                        TimeField(label: "Open", value: $day.open, enabled: canEdit && !saving)
                        Text("to")
                            .font(.footnote)
                            .foregroundStyle(.secondary)
                        TimeField(label: "Close", value: $day.close, enabled: canEdit && !saving)
                    } else {
                        Text("Closed")
                            .font(.callout)
                            .foregroundStyle(.secondary)
                        Spacer()
                    }
                }
                .padding(.vertical, 4)
            }
            if !allValid {
                ReadOnlyLine("Times are 24-hour HH:MM, and open and close can't match.")
            }
            InlineError(error)
            if canEdit {
                if dirty {
                    Button(saving ? "Saving…" : "Save hours") { save() }
                        .buttonStyle(.borderedProminent)
                        .tint(BrandColor.olive)
                        .disabled(!allValid || saving)
                        .padding(.top, 10)
                }
            } else {
                Spacer().frame(height: 4)
                ReadOnlyLine("Only owners and admins can change business hours.")
            }
        }
    }

    private func save() {
        error = nil
        saving = true
        var hours: [String: JSONValue] = [:]
        for day in days where day.enabled {
            hours[day.weekday] = .object(["open": .string(day.open), "close": .string(day.close)])
        }
        let body = JSONValue.object(["business_hours": .object(hours)])
        Task {
            do {
                let updated = try await scope.repo.updateCompany(scope.companyId, patch: body)
                onCompanyUpdated(updated)
                scope.showMessage("Business hours saved.")
            } catch {
                self.error = error.userMessage
            }
            saving = false
        }
    }
}

private struct TimeField: View {
    let label: String
    @Binding var value: String
    let enabled: Bool

    var body: some View {
        TextField(label, text: Binding(
            get: { value },
            set: { next in
                if next.count <= 5 { value = next }
            }
        ))
        .textFieldStyle(.roundedBorder)
        .font(.callout)
        .keyboardType(.numbersAndPunctuation)
        .disabled(!enabled)
        .foregroundStyle(isValidHhmm(value) ? Color.primary : BrandColor.destructive)
        .frame(maxWidth: .infinity)
    }
}

// MARK: - Away reply

private struct AwayReplyCard: View {
    let scope: SettingsScope
    let company: CompanyView
    let onCompanyUpdated: @MainActor (CompanyView) -> Void

    @State private var enabled: Bool
    @State private var message: String
    @State private var emergency: Bool
    @State private var saving = false
    @State private var error: String?

    init(scope: SettingsScope, company: CompanyView, onCompanyUpdated: @escaping @MainActor (CompanyView) -> Void) {
        self.scope = scope
        self.company = company
        self.onCompanyUpdated = onCompanyUpdated
        _enabled = State(initialValue: company.away_enabled)
        _message = State(initialValue: company.away_message ?? "")
        _emergency = State(initialValue: company.emergency_keyword_enabled)
    }

    private var canEdit: Bool { SettingsRoleGate.canEditWorkspace(scope.role) }
    private var trimmed: String { message.trimmingCharacters(in: .whitespacesAndNewlines) }
    private var dirty: Bool {
        enabled != company.away_enabled
            || trimmed != (company.away_message ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            || emergency != company.emergency_keyword_enabled
    }

    /// What actually goes out — the owner's text if they wrote one, else the
    /// product default. The preview and the #414 emergency check both read
    /// THIS, so the screen can never approve of a message that isn't sending.
    // #414 ask 5: the SERVER says what will actually send.
    private var effectiveMessage: String {
        trimmed.isEmpty ? company.away_effective_message : trimmed
    }
    private var emergencyNotice: AwayEmergencyNotice? {
        awayEmergencyNotice(emergencyEnabled: emergency, awayMessage: effectiveMessage)
    }

    var body: some View {
        SettingsCard(
            title: "Away reply",
            description: "One automatic text back when someone reaches you outside your "
                + "business hours, in your words, so you never lose an after-hours emergency."
        ) {
            LabeledToggleRow(
                label: "Reply automatically after hours",
                supporting: "Fires once per conversation when a customer first texts "
                    + "outside your hours.",
                isOn: enabled,
                enabled: canEdit && !saving
            ) { enabled = $0 }
            // The send gates refuse a US destination until the campaign is
            // approved, and the away reply is best-effort: a refusal is
            // swallowed so it never breaks inbound ingest. A switch reading ON
            // while every US customer gets silence is the first week of every
            // US workspace.
            if enabled, !usSendApproved(company) {
                ReachNote(text: usTextingOff(company)
                    ? "Customers with US numbers won't get this reply: US texting isn't on for this workspace. Canadian numbers get it now."
                    : "Customers with US numbers won't get this reply until your registration is approved. Canadian numbers get it now.")
            }
            if canEdit {
                TextField(company.away_effective_message, text: Binding(
                    get: { message },
                    set: { next in
                        if next.count <= 1000 { message = next }
                    }
                ), axis: .vertical)
                .textFieldStyle(.roundedBorder)
                .lineLimit(3 ... 8)
                .disabled(saving)
                .padding(.top, 6)
                Text("\(message.count)/1000 · {first_name} and {business_name} fill in automatically.")
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .padding(.top, 2)
            }
            // #414: the switch sits with the message that makes the offer, not
            // on a separate notifications screen. They are one decision — a
            // message inviting URGENT with the mechanism off is the exact
            // defect this issue is about, and an owner can only see it if the
            // two are on screen together.
            LabeledToggleRow(
                label: "Treat a reply of URGENT as an emergency",
                supporting: "Texts back starting with URGENT, EMERGENCY, 911 or SOS reach "
                    + "everyone on the crew straight away, at the priority that wakes a phone — "
                    + "no away reply, and never held back by your daily notification limit.",
                isOn: emergency,
                enabled: canEdit && !saving
            ) { emergency = $0 }
            // #453: which sentence appears is decided in SettingsLogic,
            // mirroring shared, so this screen, web and Android cannot drift
            // into three wordings of the same warning. Only the tone-to-colour
            // mapping is ours.
            if let notice = emergencyNotice {
                ReachNote(
                    text: notice.text,
                    tone: notice.tone == .warn ? .warn : .neutral
                )
            }
            // The preview reuses the wire's drop-empty semantics: {first_name}
            // resolves to a sample name here because the away reply DOES carry
            // the contact.
            PreviewBubble(
                label: "Preview",
                text: applyMergeFields(
                    effectiveMessage,
                    contactName: sampleFirstName,
                    businessName: company.name
                )
            )
            InlineError(error)
            if canEdit {
                if dirty {
                    Button(saving ? "Saving…" : "Save away reply") { save() }
                        .buttonStyle(.borderedProminent)
                        .tint(BrandColor.olive)
                        .disabled(saving)
                        .padding(.top, 10)
                }
            } else {
                Spacer().frame(height: 4)
                ReadOnlyLine("Only owners and admins can change the away reply.")
            }
        }
    }

    private func save() {
        if enabled && trimmed.isEmpty {
            error = "Write your away message before turning it on."
            return
        }
        error = nil
        saving = true
        let body = JSONValue.object([
            "away_enabled": .bool(enabled),
            "away_message": trimmed.isEmpty ? .null : .string(trimmed),
            "emergency_keyword_enabled": .bool(emergency),
        ])
        Task {
            do {
                let updated = try await scope.repo.updateCompany(scope.companyId, patch: body)
                onCompanyUpdated(updated)
                scope.showMessage("Away reply saved.")
            } catch {
                self.error = error.userMessage
            }
            saving = false
        }
    }
}
